src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/Marshalling/BStrStringMarshaller.cs

```diff
@@ -0,0 +1,122 @@
+    public unsafe ref struct BStrStringMarshaller
+    {
+        private byte* _ptrToFirstChar;
+        private bool _allocated;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="BStrStringMarshaller"/>.
+        /// </summary>
+        /// <param name="str">The string to marshal.</param>
+        public BStrStringMarshaller(string? str)
+            : this(str, default)
+        { }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="BStrStringMarshaller"/>.
+        /// </summary>
+        /// <param name="str">The string to marshal.</param>
+        /// <param name="buffer">Buffer that may be used for marshalling.</param>
+        /// <remarks>
+        /// The <paramref name="buffer"/> must not be movable - that is, it should not be
+        /// on the managed heap or it should be pinned.
+        /// <seealso cref="CustomTypeMarshallerFeatures.CallerAllocatedBuffer"/>
+        /// </remarks>
+        public BStrStringMarshaller(string? str, Span<byte> buffer)
+        {
+            _allocated = false;
+
+            if (str is null)
+            {
+                _ptrToFirstChar = null;
+                return;
+            }
+
+            int lengthInBytes = checked(sizeof(char) * str.Length);
+
+            // A caller provided buffer must be at least (lengthInBytes + 6) bytes
+            // in order to be constructed manually. The 6 extra bytes are 4 for byte length and 2 for wide null.
+            int manualBstrNeeds = checked(lengthInBytes + 6);
+            if (manualBstrNeeds > buffer.Length)
+            {
+                // Use precise byte count when the provided stack-allocated buffer is not sufficient
+                _ptrToFirstChar = (byte*)Marshal.AllocBSTRByteLen((uint)lengthInBytes);
+                _allocated = true;
+            }
+            else
+            {
+                // Set length and update buffer target
+                byte* pBuffer = (byte*)Unsafe.AsPointer(ref MemoryMarshal.GetReference(buffer));
+                *((uint*)pBuffer) = (uint)lengthInBytes;
+                _ptrToFirstChar = pBuffer + sizeof(uint);
+            }
+
+            // Confirm the size is properly set for the allocated BSTR.
+            Debug.Assert(lengthInBytes == Marshal.SysStringByteLen((IntPtr)_ptrToFirstChar));
+
+            // Copy characters from the managed string
+            str.CopyTo(new Span<char>(_ptrToFirstChar, str.Length));
+            ((char*)_ptrToFirstChar)[str.Length] = '\0'; // null-terminate
+        }
+
+        /// <summary>
+        /// Returns the native value representing the string.
+        /// </summary>
+        /// <remarks>
+        /// <seealso cref="CustomTypeMarshallerFeatures.TwoStageMarshalling"/>
+        /// </remarks>
+        public byte* ToNativeValue() => _ptrToFirstChar;
```

Comment:
Should this return `ushort*` and also change the type of `_ptrToFirstChar` be `ushort*`?

It is how `Utf16StringMarshaller` looks and there should look the same. Or should we switch `Utf16StringMarshaller` to the byte plan as well?

Reply:
`ushort*` on `Utf16StringMarshaller` does not inform the user about the null-terminated string contract either.

Would it be best for the `ToNativeValue`/`FromNativeValue` API to return `void*` then, on both `Utf16StringMarshaller` and `BStrStringMarshaller`?

Reply:
> Would it be best for the ToNativeValue/FromNativeValue API to use void* then, on both Utf16StringMarshaller and BStrStringMarshaller?

That is a good idea. @elinor-fung or @jkoritzinsky Thoughts?